### PR TITLE
JS: DB reads as taint sources

### PR DESCRIPTION
**Pull request by kaeluka**
This adds DB reads as taint sources in `AdditionalSources.qll`.

~~**This is not yet complete**, as it's not handling streaming database queries where the result is consumed by a pipe call (eg, knex, spanner, have such features).~~

### CODE REVIEW

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (kaeluka) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:61]: the knex model is a bit unusual in that it classifies both the expression producing the promise, as well as the `await` expression as database accesses. `getAResult`, in this case, follows the implementation of the characteristic predicate. LMK what you think.

File: javascript/ql/lib/semmle/javascript/Concepts.qll
Comment (intrigus-lgtm) [inline javascript/ql/lib/semmle/javascript/Concepts.qll:87]: Missing Ql Doc.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (intrigus-lgtm) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:676]: I think you should be able to simplify this using set-literals to
```ql
this = [database(), transaction(), snapshot()].getMember("run").getACall() and result = this.getCallback(_).getParameter(1))
```

File: javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll
Comment (intrigus-lgtm) [inline javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll:56]: Missing Ql Doc.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:676]: clever, thx! :)

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:61]: It seems reasonable.  
One thing you could do is to split up the class into two (one with the await, and one without). 
That might allow the code to be a bit cleaner.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:188]: remove those parens (and in other places)

**Comment by kaeluka** [comment]
Had a chat with @erik-krogh, I'll for now not model streaming APIs. We might choose to do that later.

**Comment by kaeluka** [comment]
left to be done:
 - implement `getResults` for rest of DBs.
 - query new results on lgtm to evaluate how noisy those sources will be.
 - bring test up to current idiom (using comments on lines that should be found).
 - performance evaluation.

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll]: I did split it up!

**Comment by kaeluka** [comment]
> * query new results on lgtm to evaluate how noisy those sources will be.

I've now used lgtm to [select all database accesses](https://lgtm.com/query/2212660374859748682/).
Octobox and strider are the projects with the highest number of database accesses. I've downloaded those locally and am running my changes against them. The result is that there's on average < 1 result per database access (because some database accesses, like data deletions or updates, do not use a result at all).

This means that there's no surprises. Does that make sense to you, @erik-krogh?

**Comment by erik-krogh** [comment]
> This means that there's no surprises. Does that make sense to you, @erik-krogh?

That makes sense to me.  
I also took at look at [your evaluation where you added these DB reads as taint sources](https://github.com/github/codeql-dca-main/tree/data/kaeluka/pr-7474-73a85c5__nightly__security-extend/reports).  
Performance seems to be within the margin of error.  

But the 4 new results all look like FPs to me.  
So I think it's the right decision to have the new source as a heuristic source.

File: javascript/ql/lib/semmle/javascript/Concepts.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/Concepts.qll:89]:
```suggestion
  DataFlow::Node getAResult() {
    none() // Overridden in subclass
  }
```

We try to keep backwards comparability with QL code written by others.  
Adding a new abstract predicate will result in a compilation error for any existing implementation. 

An alternative pattern we often use is to add the predicate with a default `none()` implementation.

File: javascript/ql/lib/semmle/javascript/Concepts.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/Concepts.qll:92]:
```suggestion
  /** 
   * Holds if the data returned is can be a user-controlled object, such as a JSON object parsed from user-controlled data. 
   * */
```
I rewrote the QLDoc. Predicates without a result should generally start with "Holds ...". 
I took some inspiration from `RemoteFlowSource::isUserControlledObject()`.

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:53]:
```suggestion
    string member;
    KnexDatabaseCallback() {
      member = ["then", "stream", "asCallback"] and
      this = knexObject().getMember(member).getACall()
    }
```
You can simplify `getAResult()` if you move `member` to a field.

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:56]:
```suggestion
      member = "then" and
```
(Belongs with the above suggestion)

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:59]:
```suggestion
      member = "asCallback" and
```
(Belongs with the above suggestion)

It also look like you need to run the autoformatter on this code (Ctrl + Shift + P -> `Format document`).

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:72]:
```suggestion
    KnexDatabaseAwait() {
      exists(AwaitExpr enclosingAwait | this = enclosingAwait.flow() | 
        enclosingAwait.getOperand() = knexObject().getAUse().asExpr()
      )
```

You have a field that is not used outside of the characteristic predicate. The field can be inlined into an `exists(..)`.

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:551]:
```suggestion
  class ExplicitQueryEvaluation extends DatabaseAccess {
```
Looks like a spurious whitespace change.  
I also think the autoformatter will fail on this.

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:559]: 2 things:  

1) You can move `DataFlow::CallNode` to the `extends` clause, that way you don't have to do an line cast here. 
2) It looks suspicious to me that you are selecting the second parameter of every callback.  
    You are saying that `foo` in `query.then((foo) => ...)` is **not** a result?

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:644]: If you add `DataFlow::CallNode` to the `extends` clause, then you don't need an inline cast here.

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:663]: If you add `DataFlow::CallNode` to the `MarsDBAccess` class, then this inline cast becomes redundant. 

There is a bunch of others that are similar, I haven't added a comment on the rest.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:191]: I think the autoformatter will fail on this.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:621]: QLDoc?

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:704]:
```suggestion
      this = v1SpannerClient().getMember("executeSql").getACall() and
      result = this.(DataFlow::CallNode).getCallback(_).getParameter(1)
```
Superfluous parentheses.

**Review by erik-krogh — COMMENTED** [review]
It's generally looking good, I just got a bunch of minor comments.  

The autoformatter is failing for a lot of your files.  

I have the following installed as a `pre-commit` hook. That hook prevents me from commiting QL code that is not formatted correctly: 

```bash
#!/bin/bash
exec 1>&2
exitVal=0
for f in $(git diff --cached --relative --name-only)
do
  filename="${f##*/}"
  extension="${filename##*.}"
  p="$PWD/$f";
  if [[ -f $p ]] && ([ $extension == "ql" ] || [ $extension == "qll" ])
  then
    codeql query format --check-only "$p"
    retVal=$?
    if [[ $retVal -ne 0 ]]
    then
      exitVal=1
    fi
  fi
done
exit $exitVal
```

File: javascript/ql/lib/semmle/javascript/Concepts.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/Concepts.qll:89]: perfect, will do

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:620]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:700]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll]: thx!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: I had no idea the autoformatter existed 🙈

File: javascript/ql/lib/semmle/javascript/Concepts.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/Concepts.qll]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:559]: 1. is locally resolved, but 2. fixed this (good catch!) and added a unit test case for it, too.

**Comment by kaeluka** [comment]
I have now fixed those issues. Thanks for the review, @erik-krogh — sorry that you had to deal with so much whitespace stuff. I now have a commit hook like your's that'll prevent this in the future :)

**Comment by erik-krogh** [comment]
> I have now fixed those issues. Thanks for the review, @erik-krogh — sorry that you had to deal with so much whitespace stuff. I now have a commit hook like your's that'll prevent this in the future :) 

Great :+1:  
Now you got a merge conflict (I think that's from Asgers MaD work).  
I'll take another look when that has been resolved.

**Comment by kaeluka** [comment]
Alright, @erik-krogh, I've rebased, resolved the conflicts, pushed.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (esbena) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:603]:
```suggestion
 * Provides classes modeling the Google Cloud Spanner library.
```
(as suggested by code scanning)

Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:603]: *sigh* I'll need a spell checker for VS Code. Thx!

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:72]: I think you're right. I tried adding `PromiseFlow::flowStep`, but it does not immediately work. I'll take a closer look.

File: javascript/ql/lib/semmle/javascript/Concepts.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/Concepts.qll:94]: 👍

**Review by esbena — COMMENTED** [review]
LGTM. But I'll defer to Erik for final approval since mine was more of a drive-by review.

**Comment by kaeluka** [comment]
Thanks, Esben! Makes sense to wait on @erik-krogh!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:605]: This entire module is mostly copy-paste from the model that was "deprecated" by models-as-data.  

You could try to use models-as-data, by extending the rows in `SpannerSinks` with new rows, where you invent some new sink-type (like `database-access`).  
In `AdditionalSources.qll` you can then access the new sinks like: `ModelOutput::getASinkNode("database-access").getARhs()`.   

Models-as-data has an `Awaited` path token that you can use. 

The rest looks fine. 

Another comment: I usually don't squash commits after un-drafting a PR. Having multiple commits makes it easier to review.

Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:605]: > You could try to use models-as-data, by extending the rows in SpannerSinks with new rows, where you invent some new sink-type (like database-access).

You mean create a new `SpannerSources` module, right? I've been looking at the MaD work (the details of which I didn't yet know all that well), and it looks promising. I'll start working on that now. Thank you.

> Another comment: I usually don't squash commits after un-drafting a PR. Having multiple commits makes it easier to review.

Thanks for pointing that out, I'll happily match what others are doing :) Is it customary to do a final squash before merging?

Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:605]: > You mean create a new `SpannerSources` module, right? I've been looking at the MaD work (the details of which I didn't yet know all that well), and it looks promising. I'll start working on that now. Thank you.

Yes, right. 
Probably just a new class `class SpannerSources extends ModelInput::SourceModelCsv`. 
And then use it like: `ModelOutput::getASourceNode("database-access").getAnImmediateUse()`. 

> Thanks for pointing that out, I'll happily match what others are doing :) Is it customary to do a final squash before merging? 

No it's not. 
The GitHub web UI shows a link from a commit to the PR that included the PR.  
That feature makes squash commits somewhat redundant.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:578]: This line had been overlooked before. Added it.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: I just pushed commit 09a28c428c25714b1b10cd4312c5e989152833e6. Looks good to you?

> No it's not.
The GitHub web UI shows a link from a commit to the PR that included the PR.
That feature makes squash commits somewhat redundant.

Alright, will do :) Thanks for pointing that out.

File: javascript/ql/test/library-tests/Security/heuristics/sources.js
Comment (erik-krogh) [inline javascript/ql/test/library-tests/Security/heuristics/sources.js:156]: Why has this line been added to the test? 
Was it to make the code in the test more realistic? (a perfectly valid reason).

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:618]:
```suggestion
        "@google-cloud/spanner;" + this.spannerClass() + ";" + this.resultPath() +
          ";database-access-result"
```
Fix code-scanning recommendation, (and an attempt to keep the auto-formatter happy).

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:446]: A `DataFlow::valueNode` where the `astNode` is of type `TaggedTemplateExpr` is always a `DataFlow::SourceNode` (see the `DefaultRange` class in `Sources.qll`).  
So you should be able to add `DataFlow::SourceNode` on top of the other classes in the extends clause. 

(Just to get rid of the inline cast to `DataFlow::SourceNode`).

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:527]: Now that you've converted the `Spanner` model to use MaD, could you do the same with your `sequelize` model?

**Review by erik-krogh — COMMENTED** [review]
We are in the final stretch, almost there :tada: 

But I still have some comments.

File: javascript/ql/test/library-tests/Security/heuristics/sources.js
Comment (kaeluka) [reply javascript/ql/test/library-tests/Security/heuristics/sources.js:156]: yes, realistic!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:618]: have done so :)

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: done!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: the second time around, it was much easier :)

File: javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll:72]:
```suggestion
    this = ModelOutput::getASourceNode("database-access-result").getAUse()
    or
    exists(DatabaseAccess dba | this = dba.getAResult())
```

One last comment, couldn't you combine the two new classes in `AdditionalSources.qll`?

Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll]: sure, done

**Review by erik-krogh — APPROVED** [review]
:tada:

**Comment by kaeluka** [comment]
Thanks for the super helpful reviews, @erik-krogh!

**Pull request merged**
